stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -1415,3 +1415,50 @@ function normalize(v::AbstractVector, p::Real = 2)
         return T[]
     end
 end
+
+# Iterator for the rows() and cols()
+struct MatrixDimIterator{T}
+    row::Bool
+    length::Int
+    M::Matrix{T}
+end
+
+function Base.iterate(it::MatrixDimIterator, state = 1)
+    if state > it.length
+        return nothing
+    elseif it.row == true
+        obj = @view it.M[state, :]
+        return (obj, state+1)
+    else
+        obj = @view it.M[:, state]
+        return (obj, state+1)
+    end
+end
+
+Base.eltype(it::MatrixDimIterator) = it.row == true ? typeof(@view it.M[1, :]) : typeof(@view it.M[:, 1])
+Base.length(it::MatrixDimIterator) = it.length
+"""
+    eachrow(M::Matrix)
+
+Get an iterator for the rows of M.
+See also [`eachcol`](@ref) and [`eachslice`](@ref).
+"""
+eachrow(M::Matrix) = MatrixDimIterator(true, size(M, 1), M)
```

Comment:
With the above change, these functions could accept all `AbstractMatrix`es.

Reply:
Thanks @mbauman for feedback. So do you think it's worth iterating on this to incorporate the changes? Or just closing the PR and waiting to merge in something from a package

Reply:
I'd suggest just simplifying this to the generator definitions I wrote above (https://github.com/JuliaLang/julia/pull/29749#issuecomment-432015741), but let's see what others think first.

Reply:
Sounds good. Committed the changes, with a clarification that `eachrow` and `eachcol` are only written for `AbstractVecOrMat`.